cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java

```diff
@@ -0,0 +1,148 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.tools;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.table.OrderedTable;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.dataset2.tx.DatasetContext;
+import co.cask.cdap.data2.dataset2.tx.Transactional;
+import co.cask.cdap.logging.save.LogSaverTableUtil;
+import co.cask.cdap.logging.write.FileMetaDataManager;
+import co.cask.tephra.DefaultTransactionExecutor;
+import co.cask.tephra.TransactionAware;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.inject.Injector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Map;
+
+/**
+ * Upgrades the Logs
+ */
+public class LogUpgrade extends AbstractUpgrade implements Upgrade {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogUpgrade.class);
+  private static final String[] OTHERS = {"logs", Constants.SYSTEM_NAMESPACE, Constants.DEFAULT_NAMESPACE};
+
+  private final Transactional<DatasetContext<OrderedTable>, OrderedTable> logMDS;
+
+  public LogUpgrade() {
+    this.logMDS = Transactional.of(
+      new TransactionExecutorFactory() {
+        @Override
+        public TransactionExecutor createExecutor(Iterable<TransactionAware> txAwares) {
+          return new DefaultTransactionExecutor(txClient, txAwares);
+        }
+      },
+      new Supplier<DatasetContext<OrderedTable>>() {
+        @Override
+        public DatasetContext<OrderedTable> get() {
+          try {
+            return DatasetContext.of(new LogSaverTableUtil(nonNamespaedFramework, cConf).getMetaTable());
+          } catch (Exception e) {
+            throw Throwables.propagate(e);
+          }
+        }
+      });
+  }
+
+  @Override
+  public void upgrade(Injector injector) throws Exception {
+    final FileMetaDataManager fileMetaDataManager = new FileMetaDataManager(new LogSaverTableUtil(
+      nonNamespaedFramework, cConf), txClient, locationFactory);
+    logMDS.execute(new TransactionExecutor.Function<DatasetContext<OrderedTable>, Void>() {
+      @Override
+      public Void apply(DatasetContext<OrderedTable> ctx) throws Exception {
+        Scanner rows = ctx.get().scan(null, null);
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(Arrays.copyOfRange(row.getRow(), FileMetaDataManager.ROW_KEY_PREFIX.length,
```

Comment:
toString() has a version with offset and length, you don't need to use Arrays.copyOfRange()